src/libsyntax/parse/parser.rs

```diff
@@ -5364,10 +5424,18 @@ impl<'a> Parser<'a> {
     /// - `const unsafe fn`
     /// - `extern fn`
     /// - etc
-    fn parse_fn_front_matter(&mut self) -> PResult<'a, (Spanned<Constness>, Unsafety, Abi)> {
+    fn parse_fn_front_matter(&mut self)
+        -> PResult<'a, (
+            Spanned<Constness>,
+            Unsafety,
+            IsAsync,
+            Abi
+        )>
+    {
         let is_const_fn = self.eat_keyword(keywords::Const);
         let const_span = self.prev_span;
         let unsafety = self.parse_unsafety();
+        let asyncness = self.parse_asyncness();
```

Comment:
So the syntax is fixed-order `default const unsafe async extern "C" fn foo() {}` now.

Fascinating.

There are two groups of things here - stuff inherent to the fn type (`unsafe`, `extern "ABI"` and `fn` itself) and possibly context-dependent "modifiers" (`default`, `const`, `async`).
So far the logic was that modifiers were grouped together on the left, and type-things were grouped together with `fn` on the right - `default const | unsafe extern "C" fn` so the modifiers can't get "inside the fn type" `type F = unsafe extern "C" fn();`.

I think we should turn the syntax into `(default const async in arbitrary order) | (unsafe extern "C" in arbitrary order) fn` eventually, but for now let's keep it fixed-order while keeping the grouping rule - `default const async | unsafe extern "C" fn`.